static/src/javascripts/projects/commercial/modules/dfp/render-advert.ts

```diff
@@ -0,0 +1,294 @@
+
+sizeCallbacks[adSizes.outstreamMobile.toString()] = (_: any, advert: Advert) =>
+	fastdom.mutate(() => {
+		advert.updateExtraSlotClasses('ad-slot--outstream');
+	});
+
+sizeCallbacks[adSizes.googleCard.toString()] = (_: any, advert: Advert) =>
+	fastdom.mutate(() => {
+		advert.updateExtraSlotClasses('ad-slot--gc');
+	});
+
+/**
+ * Out of page adverts - creatives that aren't directly shown on the page - need to be hidden,
+ * and their containers closed up.
+ */
+const outOfPageCallback = (
+	event: {
+		slot: {
+			getOutOfPage: () => unknown;
+		};
+	},
+	advert: Advert,
+) => {
+	if (!event.slot.getOutOfPage()) {
+		const parent = advert.node.parentNode as HTMLElement;
+		return fastdom.mutate(() => {
+			advert.node.classList.add('u-h');
+			// if in a slice, add the 'no mpu' class
+			if (parent.classList.contains('fc-slice__item--mpu-candidate')) {
+				parent.classList.add('fc-slice__item--no-mpu');
+			}
+		});
+	}
+	return Promise.resolve();
+};
+sizeCallbacks[adSizes.outOfPage.toString()] = outOfPageCallback;
+sizeCallbacks[adSizes.empty.toString()] = outOfPageCallback;
+
+/**
+ * Portrait adverts exclude the locally-most-popular widget
+ */
+// Temporary definition until 'geo-most-popular' is converted to TypeScript
+
+type WrappedElem = {
+	elem: HTMLElement | null;
+	remove: () => void;
+};
+sizeCallbacks[adSizes.portrait.toString()] = () =>
+	// remove geo most popular
+	geoMostPopular.whenRendered.then(
+		(popular: WrappedElem | undefined | null) =>
+			fastdom.mutate(() => {
+				if (popular?.elem) {
+					popular.elem.remove();
+					popular.elem = null;
+				}
+			}),
+	);
+
+/**
+ * Commercial components with merch sizing get fluid-250 styling
+ */
+sizeCallbacks[adSizes.merchandising.toString()] = addFluid250([
+	'ad-slot--commercial-component',
+]);
+
+const addContentClass = (adSlotNode: HTMLElement) => {
+	const adSlotContent = $$('> div:not(.ad-slot__label)', adSlotNode).get();
+
+	if (adSlotContent.length) {
+		void fastdom.mutate(() => {
+			adSlotContent[0].classList.add('ad-slot__content');
+		});
+	}
+};
+
+/**
+ * @param advert - as defined in commercial/modules/dfp/Advert
+ * @param slotRenderEndedEvent - GPT slotRenderEndedEvent
+ * @returns {Promise} - resolves once all necessary rendering is queued up
+ */
+export const renderAdvert = (
+	advert: Advert,
+	slotRenderEndedEvent: Event,
+): Promise<any> => {
+	addContentClass(advert.node);
+
+	return getAdIframe(advert.node)
+		.then((isRendered) => {
+			const callSizeCallback = (): Promise<any> => {
+				if (advert.size) {
+					let size = advert.size.toString();
+
+					if (size === '0,0') {
+						size = 'fluid';
+					}
+
+					/**
+					 * we reset hasPrebidSize to the default
+					 * value of false for subsequent ad refreshes
+					 * as they may not be prebid ads.
+					 * */
+					advert.hasPrebidSize = false;
+
+					return Promise.resolve(
+						sizeCallbacks[size]
+							? // @ts-expect-error - We check if sizeCallbacks[size] is truthy above
```

Comment:
Would `typeof sizeCallbacks[size] === 'function'` make TS happier?

Reply:
So without that expect-error we get `Cannot invoke an object which is possibly 'undefined'` on line 262:
```sizeCallbacks[size](slotRenderEndedEvent, advert)```
I feel this is erroneous because we have a truthy check for `sizeCallbacks[size]`

Reply:
Alrighty, then!